### ISSUE TITLE
[$500] Web - Chat - You can't scroll the PDF right after opening it with the "down arrow" button

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#34816]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.28.0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause - Internal Team
**Slack conversation:**

## Action Performed:
1. Navigate to a chat
2. Upload a PDF
3. Open it
4. Press the "down arrow" keyboard button

## Expected Result:
I should be able to use the arrows without clicking

## Actual Result:
You can't scroll the PDF right after opening it with the "down arrow" button. You need to click once for the function to work

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/Expensify/App/assets/78819774/c8f29f90-636c-45fa-b250-aaa508296136

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~011e97d8eae76876e6</li>
        <li>Upwork Job ID: 1748379977514749952</li>
        <li>Last Price Increase: 2024-01-26</li>
<li>Automatic offers: </li>
<ul>
<li>hungvu193 | Reviewer | 0</li>
<li>abzokhattab | Contributor | 0</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @JmillsExpensify</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @JmillsExpensify</details>

**Comment by abzokhattab** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
You can't scroll the PDF right after opening it with the down or up  buttons
### What is the root cause of that problem?
the document is focused once it's opened  

### What changes do you think we should make in order to solve the problem?

we need to call the focus function [here](https://github.com/Expensify/App/blob/c6208fe53a902a83bb49c7448f75874bda6cc682/src/components/PDFView/index.js#L142-L147): 
```js
        ref.focus();

```

**Comment by JeremyCroff** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
A newly opened PDF cannot be scrolled with keyboard without clicking on it first

### What is the root cause of that problem?
No focus is being set on the document

### What changes do you think we should make in order to solve the problem?
react-pdf has built in props to handle this.
```

 <PDFViewer innerRef={node => (this.iframe = node)}>
          <Document onRender={() => this.iframe.focus()}>
          //PDF
          </Document>
</PDFViewer>
```

I will test the focus on a password protected document as well to ensure there's no conflicts.

### What alternative solutions did you explore? (Optional)

Alternatively, you can set your focus by creating your own refs in React.

**Comment by sofi-a** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Can not scroll a PDF attachment using arrow keys.

### What is the root cause of that problem?
The document is out of focus and the arrow key presses won't register.

<img width="1728" alt="Screenshot 2024-01-21 at 14 33 32" src="https://github.com/Expensify/App/assets/31987492/380c056e-7f9f-408c-b8dc-70b2e9ed738b">

The blue border around the window appears when pressing the up/down arrow keys, showing the document is out of focus.

### What changes do you think we should make in order to solve the problem?
To focus the document after it's loaded, the following changes should be made to the [PDFView](https://github.com/Expensify/App/blob/2a03320a054f0d9a28cffc1ee96fdde78700218a/src/components/PDFView/index.js) component.

In the [`setListAttributes`](https://github.com/Expensify/App/blob/2a03320a054f0d9a28cffc1ee96fdde78700218a/src/components/PDFView/index.js#L111) function, apply the following changes

```js
setListAttributes(ref) {
...        

  setTimeout(() => ref?.focus(), 1000);

...
```

**NB: Focus won't work without the timer.**

### Result

https://github.com/Expensify/App/assets/31987492/8ba33ea6-abbf-4fba-a211-ec4f9b9b5872

**Comment by sofi-a** [comment]
@abzokhattab It didn't work for me without the timer and I set the duration randomly. It can be decreased if necessary.

**Comment by sofi-a** [comment]
@abzokhattab you're right, calling `ref.focus()` after the `ref.tabIndex = -1` statement works 👍.

**Comment by sofi-a** [comment]
I'll remove my proposal since it's redundant.

**Comment by alexpensify** [comment]
@situchan - any update here regarding these proposals?

**Comment by alexpensify** [comment]
test

**Comment by alexpensify** [comment]
test

**Comment by situchan** [comment]
reviewing

**Comment by jeremy-croff** [comment]
My proposal is currenly hidden because my github account is currently suspended while it's being recovered. The proposal should be in an email history.

The proposal was to use the react-pdf built in props to maintain focus

 <PDFViewer innerRef={node => (this.iframe = node)}>
          <Document onRender={() => this.iframe.focus()}>
            // PDF implementation
          </Document>
        </PDFViewer>

**Comment by alexpensify** [comment]
Sounds like a plan, keep us posted @situchan. Thanks!

**Comment by sourcecodedeveloper** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
After opening the pdf, the page cannot be scrolled using the down arrow key without clicking on the page view.

### What is the root cause of that problem?
The react pdf page component does not have tabIndex set and also needs to focused.

### What changes do you think we should make in order to solve the problem?
Need to set tabIndex=0 on react-pdf_Page element ref and then set focus.
https://github.com/Expensify/App/blob/0ebc6e49eb4f7217692d4d0cd0ded7c59f490253/src/components/PDFView/WebPDFPageRenderer.js#L38

```
inputRef={(el)=>{
    if(el){
        el.tabIndex=0;
        el.focus();
    }
}}
```

### What alternative solutions did you explore? (Optional)

**Comment by sourcecodedeveloper** [comment]
@situchan Please review [Proposal](https://github.com/Expensify/App/issues/34816#issuecomment-1910709631). The approach is to make the div tag hosting the react pdf page focusable by setting tabIndex = 0, by default div is not focusable and then calling focus() on the element ref.

**Comment by situchan** [comment]
I think we should hold this for:
- https://github.com/Expensify/App/pull/32800

**Comment by alexpensify** [comment]
Thanks, moving this to weekly and putting it on hold.

**Comment by situchan** [comment]
Still holding

**Comment by alexpensify** [comment]
🤦🏼 looks like my tags didn't stick last time. Moving this to Weekly!

**Comment by alexpensify** [comment]
Weekly update: On Hold

**Comment by alexpensify** [comment]
Weekly Update: On Hold

**Comment by alexpensify** [comment]
Weekly Update: On Hold

**Comment by alexpensify** [comment]
Weekly Update: @situchan it looks like #32800 is going to close. Should we close this one too or take this one off hold?

**Comment by alexpensify** [comment]
@situchan can you please update here if we should close this one too? Thanks!

**Comment by situchan** [comment]
The holding PR was closed but the root cause still remains.
There's another attempt to use focus trap in https://github.com/Expensify/App/issues/36476.

**Comment by alexpensify** [comment]
@situchan - should we hold for https://github.com/Expensify/App/issues/36476 or update the focus like @suneox suggested? Thanks!

**Comment by abzokhattab** [comment]
Just clarify [my proposal](https://github.com/Expensify/App/issues/34816#issuecomment-1900785793),  [this line](https://github.com/Expensify/App/blob/68e997d563d659cf9b5393ccb81f5c291f20a047/src/components/PDFView/index.js#L121) inside the `setListAttributes` function is gonna be reached only when the document is loaded successfully,  

as per the source code of the [Document](https://github.com/wojtekmaj/react-pdf/blob/93b09c3b0a4a231d401c5aa3f8f375f746d5c0d2/packages/react-pdf/src/Document.tsx#L609C1-L629C4) component in the `react-pdf` library, the children components will be only rendered when the pdf is loaded successfully which means the `setListAttributes` should be called after the doc is loaded.

**Comment by abzokhattab** [comment]
Correct but the `List` [component](https://github.com/Expensify/App/blob/bc3705e61c29e974161287b250a5f76f4f5d4d66/src/components/PDFView/WebPDFDocument.js#L108C1-L121C14) is only rendered when the `pageViewportsLength` is not zero and the `pageViewportsLength` is changed in the [onDocumentLoadSuccess](https://github.com/Expensify/App/blob/bc3705e61c29e974161287b250a5f76f4f5d4d66/src/components/PDFView/index.js#L85C3-L107C1) so technically speaking this line is only reached when the `List` renders which happen when the  `onDocumentLoadSuccess` is called.  

since we already update the `ref.tabIndex` in the `setListAttributes` we can call the focus at the same function. minimal code changes and fewer changes of regressions.

**Comment by alexpensify** [comment]
@situchan - when you get a chance, we need your feedback here. Thanks!

**Comment by alexpensify** [comment]
@situchan - any update here? Thanks!

**Comment by situchan** [comment]
Oh sorry this wasn't in my filtered list because of Hold in title.
I will update soon

**Comment by alexpensify** [comment]
@situchan - were you able to review this one? Thanks!

**Comment by situchan** [comment]
I will provide update in a few hrs

**Comment by situchan** [comment]
@abzokhattab's [proposal](https://github.com/Expensify/App/issues/34816#issuecomment-1900785793) looks good to me.

> since we already update the `ref.tabIndex` in the `setListAttributes` we can call the focus at the same function. minimal code changes and fewer regression chances.

This makes the most sense as `ref.tabIndex` was introduced for the exact purpose of fixing scroll issue.

🎀 👀 🎀 C+ reviewed

**Comment by situchan** [comment]
@alexpensify can we remove Hold from title and bump to Daily?

**Comment by alexpensify** [comment]
@situchan - this one has been updated, thanks for the reminder.

**Comment by abzokhattab** [comment]
Just a small update here:
we have moved all the pdf preview logic into a [new npm library](https://github.com/Expensify/react-fast-pdf). 
so the changes should be done here: https://github.com/Expensify/react-fast-pdf/blob/8592d9027094938826452b38c7010b58459c397a/src/helpers.ts#L46C1-L61C3

**Comment by abzokhattab** [comment]
a kindly reminder on this @flodnv

**Comment by mallenexpensify** [comment]
@situchan is out OOO, reassigning a C+ via the auto-assigner for fairness.

**Comment by abzokhattab** [comment]
I think we are pending on @flodnv input here, here is the last comment we have from @situchan https://github.com/Expensify/App/issues/34816#issuecomment-2015156216

**Comment by mallenexpensify** [comment]
Yup, waiting on Flo, he's off today.

**Comment by alexpensify** [comment]
Thanks for the assist @mallenexpensify!

**Comment by flodnv** [comment]
Sorry for the delay here, something weird happened with my email filters and this got lost 😕

**Comment by abzokhattab** [comment]
[The PR](https://github.com/Expensify/react-fast-pdf/pull/18) is ready! Please have a look and let me know if you have any comments @hungvu193 @flodnv

**Comment by hungvu193** [comment]
@abzokhattab I think you will also need to create a PR to bump `react-fast-pdf` in this repo as well.

**Comment by abzokhattab** [comment]
@hungvu193 sure, there we go https://github.com/Expensify/App/pull/39578

**Comment by alexpensify** [comment]
Daily Update: The PR is moving forward

**Comment by alexpensify** [comment]
@hungvu193 can we confirm if this regression notice is accurate or an error? Thanks!

**Comment by hungvu193** [comment]
> @hungvu193 can we confirm if this regression notice is accurate or an error? Thanks!

Let me check 👀

**Comment by hungvu193** [comment]
That's true, the [PR](https://github.com/Expensify/App/pull/39578) introduces the regression, since it was reverted, can you work on a fix @abzokhattab ?

**Comment by alexpensify** [comment]
@abzokhattab any update here? Is there a new PR for the fix? If yes, please share here. Thanks!

**Comment by amyevans** [comment]
👋 Hi, just as a follow up from https://github.com/Expensify/App/issues/39833, if you don't anticipate having a patch ready for `react-fast-pdf` ready shortly (which is okay!), could you revert the original PR in `react-fast-pdf` in the interim? That way further releases of `react-fast-pdf` in App are not blocked. Thanks! 🙏

**Comment by alexpensify** [comment]
## Heads up, I will be offline until Tuesday, April 23, 2024, and will not actively watch over this GitHub during that period.

If anything urgent is needed here, please ask for help in the #expensify-open-source Slack Room-- thanks!

**Comment by alexpensify** [comment]
@hungvu193 - catching up from being OOO, can we get an update on what's remaining here? Thanks!

**Comment by hungvu193** [comment]
> @hungvu193 - catching up from being OOO, can we get an update on what's remaining here? Thanks!

The PR was reverted since it was linked to a DB and we couldn't figure out the way to fix it.

**Comment by alexpensify** [comment]
Got it, so is the issue still present, or did it self-resolve? Thanks!

**Comment by hungvu193** [comment]
> Got it, so is the issue still present, or did it self-resolve? Thanks! 

Unfortunately it's still present

**Comment by alexpensify** [comment]
Thanks for the update! When I'm back online then we can bump the Open Source room to confirm if we should close this one or take a different path. 

--- 

## Heads up, I will be offline until Tuesday, May 7, 2024, and will not actively watch over this GitHub during that period.

If anything urgent is needed here, please ask for help in the #expensify-open-source Slack Room-- thanks!

**Comment by alexpensify** [comment]
This week, I've been focused on Wave Collect resources so I'll start a discussion next week.

**Comment by alexpensify** [comment]
## Heads up, I will be offline until Tuesday, May 28, 2024, and will not actively watch over this GitHub during that period.

@JmillsExpensify—If the site is in a more stable spot, please start a discussion in the Open Source Slack room about whether we should find another solution here or close this GH. I waited to have that conversation since the focus has been on higher-priority issues, so I appreciate your help while I'm OOO. '

Here is where we are at: 

>The PR was reverted since it was linked to a DB and we couldn't figure out the way to fix it.

**Comment by JmillsExpensify** [comment]
I think we should just close it. If someone is passionate, please raise a discussion in the Open Source Slack room and re-open when we've agreed on a solution.

**Comment by hungvu193** [comment]
Since I and @abzokhattab already worked on the issue. Can we request a compensation here? @JmillsExpensify

**Comment by JmillsExpensify** [comment]
What do you think is reasonable?

**Comment by hungvu193** [comment]
I think half of the bounty.

**Comment by JmillsExpensify** [comment]
That works for me, though in case anyone happens upon this decision in the future, it's a one-off and not indicative of how we'll approach cases like this in the future. In any case:

 - Contributor: @abzokhattab $250
 - Contributor+: @hungvu193 $250

**Comment by JmillsExpensify** [comment]
Contracts are paid out via Upwork. @abzokhattab also please check the message I sent you in Upwork, thanks!

**Comment by abzokhattab** [comment]
Alright thanks @JmillsExpensify .. i have refunded 250$ since i was already paid few days ago. please let me know if you have any other requests
